Restore enrollments from other nodes

linked0 opened the pull request:
Added the API endpoint called `getEnrollments` to get missing enrollments from other nodes and make the network manager periodically call the function on other nodes to add received enrollments into the enrollment pool.

Fixes #755

linked0 committed:
Get enrollments from other nodes

A node requests enrollments from other nodes only when there were
blocks received from other nodes. We do not request enrollments
from other nodes every time for network efficiency.

linked0 commented on source/agora/network/Manager.d:
```diff
@@ -1063,6 +1066,39 @@ public class NetworkManager
+        scope UTXOFinder finder) @safe nothrow
+    {
+        // TODO: This should be part of the configuration.
+        static uint MinNodes = 3;
```
I thought that getting enrollments from all other nodes could impose a burden on the network. So I'm sampling the nodes which will get the request, or `getEnrollments`. @Geod24  How do you think of this approach?

Geod24 replied:
Getting all enrollments also means that we most likely always have needless data transfer even when we're up to date, which isn't optimal.

linked0 replied:
OK, I see what you mean. I will find a way to store missing enrollments that will be requested later from other nodes.

linked0 replied:
Had a different approach and solved it.

Geod24 commented on source/agora/consensus/data/Enrollment.d:
```diff
@@ -70,6 +70,21 @@ public struct Enrollment
     }
 }
 
+/*******************************************************************************
+
+    Define enrollment stored in the pool
+
+*******************************************************************************/
+
+public struct PoolEnrollment
+{
+    /// An enrollment data in the pool
+    public Enrollment enrollment;
+
+    /// The available height set on being stored in the pool
+    public Height avail_height;
+}
```
I don't follow why we need to get the `avail_height` from other nodes ?

linked0 replied:
The `avail_height` is needed to store enrollments into the pool because a requesting node could have a different height and can not set the exact value for the field.

linked0 replied:
Anyway, I'll look into more whether the field could be removed.

linked0 replied:
I think it's needed. So I remain the field as it is.

Geod24 replied:
The reason I'm asking is because we cannot trust the remote node to provide the right information.
What if the node sends us a right enrollment but pretends it will be available in 3 million blocks ?

linked0 replied:
After considering your opinion. I could use a new table for missing enrollments that have `Enrollment` and `avail_height`. How do you think of creating new table for that? I mean nodes can just communicate with `Enrollment`, not `PoolEnrollment`.

linked0 replied:
Anyway, I will find a way to use `Enrollment` in the API endpoint, `getEnrollments`, and remove this structure.

Geod24 commented on source/agora/api/FullNode.d:
```diff
@@ -359,4 +359,17 @@ public interface API
+
+    ***************************************************************************/
+
+    @method(HTTPMethod.GET)
```
This is the default

linked0 replied:
Removed

Geod24 commented on source/agora/test/GetEnrollments.d:
```diff
@@ -0,0 +1,80 @@
+module agora.test.GetEnrollments;
```
DDOC

linked0 replied:
Added DDOC

Geod24 commented on source/agora/node/FullNode.d:
```diff
@@ -500,6 +500,7 @@ public class FullNode : API
         {
             log.error("Error sending updated block headers:{}", e);
         }
+        this.network.getEnrollments(this.enroll_man, this.utxo_set.getUTXOFinder());
```
I think it should be its own task, so that it doesn't starve because of block catchup

linked0 replied:
Modified.

linked0 committed:
Make the catchup period for missing enrollments configurable

We can now set duration between requests for retrieving the missing
enrollments from all other nodes.

linked0 commented:
Ready for review

hewison-chris commented on source/agora/node/Config.d:
```diff
@@ -224,6 +224,10 @@ public struct NodeConfig
     /// from all other nodes
     public Duration block_catchup_interval = 20.seconds;
 
+    /// The number of the other nodes from which a node can retrive missing
```
I think you need to update this comment.

linked0 replied:
Oh, I missed it. Will update.

linked0 replied:
Updated.

hewison-chris commented on source/agora/node/FullNode.d:
```diff
@@ -1043,6 +1043,19 @@ public class FullNode : API
         return this.enroll_man.getEnrollment(enroll_hash);
     }
 
+    /// GET: /enrollments
+    public override PoolEnrollment[] getEnrollments (Set!Hash enroll_keys) @safe
```
Please add to source/api/spec.yaml.

linked0 replied:
Will update.

linked0 replied:
Added.

hewison-chris commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -111,6 +112,9 @@ public class EnrollmentManager
     /// Ditto
     private PreImageCycle cycle;
 
+    /// Missing enrollments appearing from a failure in adding an enrollment
```
```suggestion
    /// Missing enrollments discovered from other nodes
```

linked0 replied:
Will update.

hewison-chris commented on source/agora/test/GetEnrollments.d:
```diff
@@ -0,0 +1,138 @@
+    CustomAPIManager.enable_catchup = true;
+    network.expectHeight([GenesisValidators - 1], Height(GenesisValidatorCycle - 1));
+
+    Thread.sleep(conf.node.enrollment_catchup_interval);
```
How long will this sleep? I think the default is too long for the test and should be overridden to maybe 500.msecs?

hewison-chris replied:
Also I think it is better to use retry with timeout rather than sleep.

linked0 replied:
I will try to use retry.

linked0 replied:
Used retry. It's better.

hewison-chris reviewed (changes requested):
Looks good just a few requests.

hewison-chris commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -111,6 +112,9 @@ public class EnrollmentManager
     /// Ditto
     private PreImageCycle cycle;
 
+    /// Missing enrollments appearing from a failure in adding an enrollment
```
```suggestion
    /// Set to store `utxo hashes` of enrollments that fail to be added when initially received
```

linked0 replied:
Modified with a shorter message.

hewison-chris commented on source/agora/test/Base.d:
```diff
@@ -2006,6 +2006,9 @@ public struct TestConf
         block_catchup_interval: 1.seconds,
 
+        // Catchup period for missing enrollments
+        enrollment_catchup_interval: 1.seconds,
```
👍

hewison-chris commented on source/agora/test/GetEnrollments.d:
```diff
@@ -0,0 +1,137 @@
+
+    // check the missing enrollment to be restored
+    retryFor(bad_node.getEnrollment(enroll.utxo_key) == enroll,
+        5.seconds, format!"The enrollment (%s) not in pool of the bad validator"
```
```suggestion
        2 * conf.node.enrollment_catchup_interval, format!"The enrollment (%s) not in pool of the bad validator"
```

hewison-chris replied:
It should happen within 2 intervals.

linked0 replied:
Updated. Ty.

Geod24 commented on source/agora/consensus/pool/Enrollment.d:
```diff
@@ -17,6 +17,7 @@
 module agora.consensus.pool.Enrollment;
 
 import agora.common.ManagedDatabase;
+import agora.common.Set;
```
Obviously this is not used in this commit.

linked0 replied:
Removed.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -738,6 +771,42 @@ public class EnrollmentManager
+
+    ***************************************************************************/
+
+    public Set!Hash getMissingEnrollments () @safe
+    {
+        Set!Hash local_missing_enrolls;
+        foreach (key; this.missing_enrolls)
+        {
+            if (this.getEnrollment(key) == Enrollment.init)
+                local_missing_enrolls.put(key);
+        }
+
+        this.missing_enrolls = local_missing_enrolls;
+        return missing_enrolls;
+    }
```
Why the `Set` duplication ?

linked0 replied:
The `missing_enrollments` will be used later so it must have the latest missing enrollments. If you meant another way, please let me know.

Geod24 replied:
I am not sure why we need a local `Set` for this. This seems to replicate what was done with "missing transactions", but the problem was different: Missing transactions uses a `Set` because we see the hashes in the SCP statement but we can't match a data structure. On the other hand, we haven't seen the hashes of `Enrollment` that are missing, we don't even know if they exist or not.

linked0 replied:
I see. I think we need to store `Enrollment`s for the missing enrollments instead of hashes as commented on other threads. For `Set` duplication, I didn't know that `remove` function can be used in the iteration. I understand the local variable is not needed now. Thx for pointing out the problem.

Geod24 commented on source/agora/node/FullNode.d:
```diff
@@ -1026,7 +1043,12 @@ public class FullNode : API
-        this.utxo_set.peekUTXO(enroll.utxo_key, utxo);
+        if (!this.utxo_set.peekUTXO(enroll.utxo_key, utxo))
+        {
+            this.enroll_man.addMissingEnrollment(enroll.utxo_key);
```
Wait that is not a "missing enrollment", it's a missing UTXO referenced in an enrollment

linked0 replied:
I see what you mean. I will try to save missing `Enrollment`s instead of UTXOs.

Geod24 commented on source/agora/node/FullNode.d:
```diff
@@ -502,6 +504,21 @@ public class FullNode : API
+        if (this.network.peers.empty())  // no clients yet (discovery)
+            return;
+
+        this.network.getEnrollments(this.enroll_man, this.utxo_set.getUTXOFinder());
```
This looks very dangerous. We can't really use the UTXO finder in tasks as it is not re-entrant.
Meaning that the following may occur:
```
Task A is called, starts to use the UTXOFinder, and yield mid-task;
Task B is called, starts to use the UTXOFinder, which resets the list of double spend, and yield mid-task;
Task A resumes, continues using the UTXOFinder, unaware that it lost replay protection.
```

linked0 replied:
Right, I missed it, I will use peekUTXO instead.

linked0 replied:
Updated. It uses `peekUTXO` now.

Geod24 commented on source/agora/network/Manager.d:
```diff
@@ -1063,6 +1066,44 @@ public class NetworkManager
+        scope UTXOFinder finder) @safe nothrow
+    {
+        NodeConnInfo[] conns = this.peers.array;
+        ulong[] conn_indices = conns.length.iota.array;
```
So this essentially creates an array being `[0, 1, 2, 3, ...]` ?
And it *only* uses it for a `foreach`. We should try to avoid pointless memory allocations like this.
A simple way to do this is:
```D
foreach (idx; 0 .. conns.length)
```

But then, the only use of `idx` in that loop is for `conns[idx]`, so why not simply:
```D
foreach (conn; this.peers)
```

And avoid two superfluous memory allocations ?

linked0 replied:
Updated.

Geod24 commented on source/agora/network/Manager.d:
```diff
@@ -1063,6 +1066,44 @@ public class NetworkManager
+            auto pool_enrolls = conns[idx].client.getEnrollments(missing_enrolls);
+            foreach (enroll; pool_enrolls)
+            {
+                import std.stdio;
```
Leftover

linked0 replied:
Removed.

Geod24 commented on source/agora/test/GetEnrollments.d:
```diff
@@ -0,0 +1,137 @@
+        }
+
+        ///
+        protected override void catchupTask () nothrow
+        {
+            if (atomicLoad(*this.enable_catchup))
+                super.catchupTask();
+        }
```
Why do we override `catchupTask` and not `catchupEnrollment` ?

linked0 replied:
It's because I wanna make the bad node one block behind the latest height.

Geod24 reviewed (changes requested):
I don't understand the logic of the last commit. This is not catching up `Enrollment`, since the only place where `addMissingEnrollment` is called is when we receive an `Enrollment` that we don't have an UTXO for. So when we do so, we'll reject it, and ask other nodes to send it to us again... What am I missing ?

linked0 commented:
I applied a table for orphan enrollments and pushed the code in this PR, but @Geod24 gave an opinion that it's not in fact needed IRL.
@Geod24 Could you explain it in more detail?

linked0 commented:
@Geod24 Do you totally disagree with this approach? Could you review this once?

linked0 commented on source/agora/consensus/pool/Enrollment.d:
```diff
@@ -191,6 +245,28 @@ public class EnrollmentPool
+
+    ***************************************************************************/
+
+    public void removeOrphan (in Hash enroll_hash) @trusted nothrow
```
Additionally, we need the function named `removeOrphanExpired` or to add the logic removing orphan enrollments to the `removeExpired` function.

Geod24 commented:
From what I can see from the diff:
- Whenever we receive an `Enrollment` which does not pass validation, we store it in a table, regardless of **why** is wasn't valid (https://github.com/bosagora/agora/pull/2441/files#diff-71a378ff4dee5d52c4fd906503085d5635555f0ed418e138439cb7f78dc575dbR96);
- We only remove entries from this table if we add a validated enrollment (https://github.com/bosagora/agora/pull/2441/files#diff-71a378ff4dee5d52c4fd906503085d5635555f0ed418e138439cb7f78dc575dbR153);
- Although, we also override data if the same UTXO is used (https://github.com/bosagora/agora/pull/2441/files#diff-71a378ff4dee5d52c4fd906503085d5635555f0ed418e138439cb7f78dc575dbR188-R192);
- We have a periodic table which allocates a whole AA with the content of this database (https://github.com/bosagora/agora/pull/2441/files#diff-7b4de96b302d2ab907d0084dd3697c8dbd7ea2359812b46ec950c0e3c94a17c7R1077);
- We request all our peers for the enrollments we already know about (???): https://github.com/bosagora/agora/pull/2441/files#diff-7b4de96b302d2ab907d0084dd3697c8dbd7ea2359812b46ec950c0e3c94a17c7R1091

Provided the above is correct, there are many problems that would ensue from this:
1) First and foremost, we **know** the content of the `Enrollment` structure. Why are we asking our peers to send us something we know about ?
2) Even if there was a good reason for **1** (I can't see any), we don't take into account **why** validation fails. It could be a wrong signature, in which case, there's no point in storing it;
3) Since we don't take into account the failure reason, this is a glaring DoS vector: One can just send a random piece of data with `utxo_key == 0x00...001`, then `0x.00...002`, etc... and we will store **everything** in the database. We will then quickly run of memory because of the AA allocations, or of disk space because we essentially would store `(Enrollment.sizeof +sqlite3 overhead) * 2^(64-1)`;

linked0 commented:
> From what I can see from the diff:
> ...

I agree with your opinion. Sorry for not thinking to that extent about that in advance. Will think about it more.

linked0 committed:
Add getEnrollments API endpoint

This endpoint is used to get enrollments from the enrollment pool of
other nodes.

linked0 commented:
Ready for review.
I applied another approach which is not perfect but better than the current status. Getting enrollments could make the network congested so I considered making the burden of the network as small as possible by getting enrollments from other nodes only when other nodes have more enrollments.

linked0 commented:
Ready for review.
This is a more improved version that is more compact and removes the unnecessary `PoolEnrollment` structure.

Geod24 commented on source/agora/node/FullNode.d:
```diff
@@ -1062,6 +1062,17 @@ public class FullNode : API
         return this.enroll_man.getEnrollment(enroll_hash);
     }
 
+    /// GET: /enrollments
+    public override Enrollment[] getEnrollments (ulong height, ulong enroll_count)
+        @safe
+    {
+        auto enrolls = this.enroll_man.enroll_pool.getEnrollments(Height(height));
+        if (enrolls.length > enroll_count)
+            return enrolls;
+        
+        return Enrollment[].init;
+    }
+
```
I don't understand this API. I don't see why it would make sense from the point of view of either the client, or the server.

In the client:
- You probably want to set this to a low value (I would just use 0?) because otherwise you might not get any result;

In the server:
- You always allocate, then check the length, then discard the allocated array if the user-provided value is too high;
- If an attacker repeatedly sends you `ulong.max`, you'll just end up allocating like crazy;

So the endpoint is an easy DoS vector, and doesn't make sense to the client either. What am I missing ?

linked0 replied:
Modified the code vulnerable to DoS attack.

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -226,6 +226,27 @@ public class EnrollmentManager
                                     &this.validator_set.findRecentEnrollment);
     }
 
+    /***************************************************************************
+
+        Add a enrollment data with the available height to the enrollment pool
+
+        Params:
+            enroll = the enrollment data to add
+            avail_height = block height for enrollment
+            finder = the delegate to find UTXOs with
+
+        Returns:
+            true if the enrollment data has been added to the enrollment pool
+
+    ***************************************************************************/
+
+    public bool addEnrollment (in Enrollment enroll, in Height avail_height,
+        scope UTXOFinder peekUTXO) @safe nothrow
+    {
+        return this.enroll_pool.add(enroll, avail_height, peekUTXO,
+                                    &this.validator_set.findRecentEnrollment);
+    }
+
```
I've been trying to remove those trampoline functions from the EM for the past year.
The EM role is to manage a validator's enrollment, not to be an interface for the ValidatorSet.
Well, not anymore, although that was the original design, it proved inadequate.

linked0 replied:
Got it. Removed it. Thx.

Geod24 commented on source/agora/network/Manager.d:
```diff
@@ -1044,6 +1056,26 @@ public class NetworkManager
+    public void getEnrollments (Height height, EnrollmentManager enroll_man,
+        scope UTXOFinder finder) @safe nothrow
+    {
+        Hash[] enroll_keys;
+        enroll_man.enroll_pool.getEnrollments(height).
+            each!(enroll => enroll_keys ~= enroll.utxo_key);
+
+        foreach (peer; this.peers)
+        {
+            auto peer_enrolls = peer.client.getEnrollments(height, enroll_keys.length);
+            foreach (enroll ; peer_enrolls)
+            {
+                if (!enroll_keys.canFind(enroll.utxo_key) && 
+                    enroll_man.addEnrollment(enroll, height, finder))
+                    enroll_keys ~= enroll.utxo_key;
+            }
+        }
```
So this code:
1) Get a (newly allocated) array of all the enrollments;
2) Allocate a new array of the same size with just the `utxo_key`;
3) Queries *all peers* (no matter what), and for each peer:
4) Get all their enrollments (but only if they have more enrollments... If they have less, but a different set, no luck);
5) Add each enrollment, after doing a linear search in `enroll_keys`;

Not only is this approach extremely wasteful, as the pool checks if there's already an enrollment for this UTXO, hence the `getEnrollments` => `each` / `canFind` are superfluous, but it also won't work if, for example, we have 5 enrollments, and the other node has 4 **totally different** Enrollments.

linked0 replied:
Updated.
But there remains the issue that needless data transfer occurs even when we're up to date, which isn't optimal. It might need a whiteboard.

Geod24 commented on source/agora/node/FullNode.d:
```diff
@@ -506,6 +506,26 @@ public class FullNode : API
             log.error("Error sending updated block headers:{}", e);
         }
+
+        // try to catch up enrollments only when there some blocks received
```
Why ?

linked0 replied:
It was for efficient networking but I took another approach.

Geod24 reviewed (changes requested):
As mentioned in the commit, this allocates way too much, but it also won't work.

linked0 commented:
Thanks for the review. Will request a whiteboard after reviewing the comments.

linked0 committed:
Add function getting UTXOs by output type

omerfirmak commented:
Is this really necessary?
Nodes already broadcast everything in their enrollment pool during the nomination phase anyway.

linked0 commented:
> Is this really necessary? Nodes already broadcast everything in their enrollment pool during the nomination phase anyway.

Yes, it is because we need to synchronize the enrollment pool as far as possible before the nomination phase.

linked0 commented:
Ready for review.

omerfirmak commented:
I still don't get it.

Enrollments are nominated as they are, so no hashing like Transactions. Even if a node has no enrollment in it's pool, it can validate and vote on other nominations.

Also, Enrollments are valid only on a single Block. So after any new Block, pools are cleared, effectively syncing the whole network.

linked0 commented:
You're right, but if there are any missing enrollments in the pool on a validator that already exist on other nodes, the validator can not make the proper candidates be validators in the next block by nominating consensus data. Am I missing something?

omerfirmak commented:
That particular node will fail to nominate those Validators, yes. But some nodes will do. Our `combineCandidates` should prioritze the nominations with enrollments over the ones with no enrollments.

I still feel like our `combineCandidates` is lacking a lot of flexibility. It should be aggregating TXs and Enrollments from all the nominations and creating an optimal one from all of those, not just picking one of the nominations like we do right now, according to a suboptimal criteria.

With all these "syncing", we are going against the nature of a distributed system. It is completely fine (and probably better, since it acts as a load balance) to have different states in different nodes, as long as we update the chain with a somewhat globally optimal set of TXs and enrollments.

linked0 commented:
> Our combineCandidates should prioritze the nominations with enrollments over the ones with no enrollments.

I think it's a good point for the situation. I will think of it and try to apply it after discussing more with @Geod24. Thanks for the good idea and please let me know if you think I'm still missing something.
